In [4]:
import numpy as np 
import pandas as pd

In [10]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv']))



In [11]:
# 158 with wc 2022 teams

In [12]:
df.shape

(182, 17)

In [13]:
df

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0   0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1   0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2   0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3   0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4   0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   
..       ...       ...        ...           ...       ...           ...   
1   0.172308  0.074615   0.113846      0.106154 -0.288462     -0.288462   
2  -0.247692 -0.058462   0.023077     -0.005385 -0.083077     -0.176154   
3   0.132308  0.074615  -0.188462     -0.361538 -0.501538     -0.310000   
4  -0.051538 -0.145385   0.025385     -0.077692  0.152308     -0.122308   
5   0.110000 -0.004615  -0.030769     -0.038462 -0.182308      0.062308   

    takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0    0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1    0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2    0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3    0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4    0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   
..        ...         ...        ...              ...           ...       ...   
1    0.229231    0.199231  -0.517692        -0.470000     -0.401538 -0.266923   
2    0.573846   -0.158462  -0.038462        -0.166923     -0.096923  0.683077   
3   -0.039231   -0.035385  -0.615385        -0.616154     -0.137692 -0.262308   
4    0.416154    0.668462   0.161538         0.006154      0.262308  0.083846   
5    0.568462   -0.147692  -0.150769        -0.020000      0.100000 -0.142308   

    aerials_z   fouls_z  fouled_z    nsxg_z  results  
0   -0.020000 -0.224615  0.100769 -0.124615        2  
1    0.196923 -0.016923  0.033846  0.146923        1  
2    0.683846  0.213846  0.353846  0.786923        9  
3    0.038462  0.039231  0.022308  0.692308        5  
4    0.572308 -0.016154 -0.096923  0.890769        5  
..        ...       ...       ...       ...      ...  
1   -0.225385 -0.511538 -0.307692  0.158462       12  
2   -0.409231 -0.192308 -0.297692  0.074615       27  
3    0.105385 -0.408462 -0.306923 -0.370769       27  
4    0.110769  0.229231 -0.048462  0.018462       16  
5   -0.058462  0.201538  0.229231 -0.026923       21  

[182 rows x 17 columns]

In [14]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.602349   0.062162      0.314568  0.241564   
xg_z             0.602349  1.000000   0.398585      0.673510  0.373094   
crosses_z        0.062162  0.398585   1.000000      0.661005  0.250992   
boxtouches_z     0.314568  0.673510   0.661005      1.000000  0.389432   
passes_z         0.241564  0.373094   0.250992      0.389432  1.000000   
progpasses_z     0.102207  0.173024   0.281983      0.238298  0.745008   
takeons_z        0.260127  0.313048   0.131417      0.203435  0.360414   
progruns_z       0.218886  0.438346   0.280219      0.437009  0.646392   
tackles_z        0.302733  0.410813   0.225715      0.412740  0.755495   
interceptions_z  0.241081  0.326098   0.215640      0.324611  0.736127   
clearances_z    -0.173962 -0.364917  -0.224268     -0.404166 -0.113648   
blocks_z         0.030472  0.080053   0.039333      0.175103  0.418245   
aerials_z        0.086392  0.084507   0.261073      0.279718  0.014513   
fouls_z         -0.146084 -0.077126   0.020417     -0.081497 -0.003338   
fouled_z         0.050683  0.188836   0.204139      0.170130  0.298358   
nsxg_z           0.373758  0.665801   0.672070      0.863156  0.511126   
results         -0.628505 -0.333427   0.019783     -0.214594 -0.172610   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.102207   0.260127    0.218886   0.302733   
xg_z                 0.173024   0.313048    0.438346   0.410813   
crosses_z            0.281983   0.131417    0.280219   0.225715   
boxtouches_z         0.238298   0.203435    0.437009   0.412740   
passes_z             0.745008   0.360414    0.646392   0.755495   
progpasses_z         1.000000   0.152443    0.397798   0.524199   
takeons_z            0.152443   1.000000    0.561462   0.349560   
progruns_z           0.397798   0.561462    1.000000   0.565384   
tackles_z            0.524199   0.349560    0.565384   1.000000   
interceptions_z      0.535788   0.329728    0.468762   0.716035   
clearances_z        -0.003754  -0.057416   -0.066807  -0.213072   
blocks_z             0.385167   0.191039    0.318510   0.281585   
aerials_z            0.040642  -0.062724    0.012057   0.098272   
fouls_z              0.169039   0.051742    0.011865  -0.045014   
fouled_z             0.157173   0.297329    0.268676   0.299043   
nsxg_z               0.374118   0.374014    0.550929   0.496909   
results             -0.083340  -0.153788   -0.215117  -0.217240   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.241081     -0.173962  0.030472   0.086392 -0.146084   
xg_z                    0.326098     -0.364917  0.080053   0.084507 -0.077126   
crosses_z               0.215640     -0.224268  0.039333   0.261073  0.020417   
boxtouches_z            0.324611     -0.404166  0.175103   0.279718 -0.081497   
passes_z                0.736127     -0.113648  0.418245   0.014513 -0.003338   
progpasses_z            0.535788     -0.003754  0.385167   0.040642  0.169039   
takeons_z               0.329728     -0.057416  0.191039  -0.062724  0.051742   
progruns_z              0.468762     -0.066807  0.318510   0.012057  0.011865   
tackles_z               0.716035     -0.213072  0.281585   0.098272 -0.045014   
interceptions_z         1.000000     -0.090627  0.334052   0.061036  0.056344   
clearances_z           -0.090627      1.000000  0.201853   0.203408  0.284711   
blocks_z                0.334052      0.201853  1.000000   0.019995  0.219356   
aerials_z               0.061036      0.203408  0.019995   1.000000  0.091970   
fouls_z                 0.056344      0.284711  0.219356   0.091970  1.000000   
fouled_z                0.338488      0.095257  0.122628   0.027483  0.213572   
nsxg_z                  0.428819     -0.388970  0.192452   0.167996 -0.031442   
results                -0.157544      0.091331 -0.068228  -0.039906  0.134421   

                 fo

In [80]:
df.shape

(158, 17)

In [92]:
# df.to_csv('WC-stats.csv', index=False)


In [82]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [83]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [84]:
y = df['results']

In [85]:
X = df.drop(columns=['results'])

In [86]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [87]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.42629810615686115

In [88]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [90]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 1.282 +/- 0.149
    progruns_z0.096 +/- 0.029
    xg_z    0.095 +/- 0.030
    takeons_z0.026 +/- 0.013
neg_mean_absolute_percentage_error
    goals_z 0.988 +/- 0.115
neg_mean_squared_error
    goals_z 108.981 +/- 12.672
    progruns_z8.139 +/- 2.494
    xg_z    8.047 +/- 2.509
    takeons_z2.195 +/- 1.067


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.